frontend/catalyst/api_extensions/callbacks.py

```diff
@@ -215,6 +216,7 @@ def __call__(self, args):
     def _check_types(self, obs, exp_aval):
         """Raise error if observed value is different than expected abstract value"""
         obs_aval = shaped_abstractify(obs)
+        obs_aval = obs_aval.strip_weak_type()
```

Comment:
What's this for? This removes the info that it is a python type right?

Reply:
No. It removes the info that a type is weak and can be promoted. https://jax.readthedocs.io/en/latest/type_promotion.html#weak-types

Reply:
Yes, that is exactly what happens with Python types (like `int`). Numpy types on the other hand are "strong" types (like `int64`).

But you didn't answer why it is being done.

Reply:
It is there because the shaped_abstractify will always add the weak_type.

Reply:
Does it? From the source it looks it applies it conditionally on the input:
https://github.com/google/jax/blob/2679ece82db68d7d7e5bf1353b2586d75229ca9a/jax/_src/api_util.py#L572
https://github.com/google/jax/blob/2679ece82db68d7d7e5bf1353b2586d75229ca9a/jax/_src/api_util.py#L613

Is there some other context I am missing?

Reply:
Also it seems like `strip_weak_type` just ... doesn't do anything?
https://github.com/google/jax/blob/2679ece82db68d7d7e5bf1353b2586d75229ca9a/jax/_src/core.py#L1393

Reply:
Yes, you are missing going into the try statement. Look into the implementation of [raise_to_shaped](https://github.com/google/jax/blob/41a1f2cfdc4dd2e05b32c57b871f0dc6aa3c26a8/jax/_src/core.py#L2052-L2061)

```
def raise_to_shaped(aval: AbstractValue, weak_type=None):
  aval_type = type(aval)
  if aval_type is ShapedArray and weak_type is None:
    return aval
  if weak_type is None:
    weak_type = getattr(aval, 'weak_type', False)
  for typ in aval_type.__mro__:
    handler = raise_to_shaped_mappings.get(typ)
    if handler: return handler(aval, weak_type)
  raise TypeError(type(aval))
```

Reply:
I stand corrected, it is not needed.

Reply:
Alright, I mean it could very well be needed I just wanted to understand why :)